evaluation/utils/shared.py

```diff
@@ -235,65 +229,58 @@ def prepare_dataset(
     """Update the progress bar and write the result to the output file."""
-    try:
-        if isinstance(result_or_future, Future):
-            result = result_or_future.result()
-        else:
-            result = result_or_future
-    except Exception as e:
-        # Handle the error
-        # Exception may be raised in the process_instance_func and will
-        # be raised here when we try to access the .result() of the future
-        handle_error(
-            EvalError(
-                instance_id=instance.instance_id,
-                error=str(e),
-                stacktrace=traceback.format_exc(),
-            ),
-            instance,
-            pbar,
-            instance_queue,
-        )
-        return
+    pbar.update(1)
+    pbar.set_description(f'Instance {result.instance_id}')
+    pbar.set_postfix_str(f'Test Result: {result.test_result}')
+    logger.info(
+        f'Finished evaluation for instance {result.instance_id}: {str(result.test_result)[:300]}...\n'
+    )
+    output_fp.write(json.dumps(result.model_dump()) + '\n')
+    output_fp.flush()
 
-    # Update the progress bar and write the result to the output file
-    if isinstance(result, EvalOutput):
-        pbar.update(1)
-        pbar.set_description(f'Instance {result.instance_id}')
-        pbar.set_postfix_str(f'Test Result: {result.test_result}')
-        logger.info(
-            f'Finished evaluation for instance {result.instance_id}: {str(result.test_result)[:300]}...\n'
-        )
-        output_fp.write(json.dumps(result.model_dump()) + '\n')
-        output_fp.flush()
-    elif isinstance(result, EvalError):
-        handle_error(result, instance, pbar, instance_queue)
-    else:
-        raise ValueError(f'Unexpected result type: {type(result)}')
 
+def _process_instance_wrapper(
+    process_instance_func: Callable[[pd.Series, EvalMetadata, bool], EvalOutput],
+    instance: pd.Series,
+    metadata: EvalMetadata,
+    use_mp: bool,
+    max_retries: int = 5,
+) -> EvalOutput:
+    """Wrap the process_instance_func to handle retries and errors.
 
-def handle_error(
-    error: EvalError, instance: pd.Series, pbar: tqdm, instance_queue: mp.Queue
-):
-    """Handle an error that occurred during evaluation."""
-    logger.error(
-        f'Retrying instance [{instance.instance_id}] due to error: {error.error}. Stacktrace:\n{error.stacktrace}'
-        + '\n'
-        + '-' * 10
-        + '[You may ignore this error if it is a transient issue - the instance will be automatically retried.]'
-        + '-' * 10
-        + '\n'
-    )
-    instance_queue.put(instance)
-    pbar.total += 1
-    pbar.refresh()
+    Retry an instance up to max_retries times if it fails (e.g., due to transient network/runtime issues).
+    """
+    for attempt in range(max_retries):
+        try:
+            result = process_instance_func(instance, metadata, use_mp)
+            return result
+        except Exception as e:
+            if attempt == max_retries - 1:
```

Comment:
Following the last comment

```suggestion
            if attempt == max_retries:
```